astropy/convolution/convolve.py

```diff
@@ -353,6 +356,9 @@ def convolve(array, kernel, boundary='fill', fill_value=0.,
         result[initially_nan] = np.nan
 
     # Convert result to original data type
+    if passed_quantity is not None:
```

Comment:
Looking at this again, I think since we keep on purpose a reference to the input array, we might as well look for the unit only here, i.e., something like
```
unit = getattr(passed_array, 'unit', None)
if unit is not None:
    result <<= unit
```
(and the same for the other routine)

Reply:
Good points! In convolve_fft a reference to the input array isn't kept, so I didn't change the positioning of the line which looks for the unit. I've changed everything else you pointed out though.